Added interception support

rajiv448 opened the pull request:
- Added logger module
- Added interception support for Linux platform (interception on only few selected APIs has been added in this commit)
- Added interception support for Windows platform.
- Extended the interception support for other APIs

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -118,6 +128,75 @@ void log_d(const Args&... args)
+ */
+bool file_exists(const std::string& path)
+{
+#ifdef _WIN32
+  DWORD file_attr = GetFileAttributes(path.c_str());
+  return !(file_attr == INVALID_FILE_ATTRIBUTES);
+#else
+  struct stat buffer{};
+  return (stat(path.c_str(), &buffer) == 0);
+#endif /* #ifdef _WIN32 */
```
Here I would recommend to use std::filesystem.  It would be same code on Windows and Linux.

rajiv448 replied:
Done, please check updated commit

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -175,15 +255,82 @@ int parse_cmdline(launcher& app, int argc, char* argv[])
 
-  split_by_ws(argv[argc-1], app.m_child_cmd_args);
-  for (auto& element : app.m_child_cmd_args)
+  if (optind <= static_cast<int>(args.size() - 1))
```
Shouldn't the cast be on the other size?  `static_cast<size_t>(optind)`  smaller to larger?

rajiv448 replied:
Correct, fixed in update commit

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -175,15 +255,82 @@ int parse_cmdline(launcher& app, int argc, char* argv[])
+  putenv_t(app.m_app_name_env);
+
+  /* Capture current time */
+  struct timespec start_time = {0, 0};
+  clock_gettime(CLOCK_REALTIME, &start_time);
```
It would be better if  `<chrono>` could be used.

rajiv448 replied:
As suggested, used chrono for reading time. Please check latest update and let me know if any further changes need to be made.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+  xrt_device_inst.cpp
+)
+
+if (WIN32)
+  add_definitions( -DXCL_DRIVER_DLL_EXPORT )
+  add_definitions( -DXRT_API_SOURCE )
+endif()
```
These should never be defined in make files.  You are setting these defines only if your compiled object code is in the same library that exports API sources, e.g. in xrt_coreutil.  See any file in core/common/api/*.cpp and how these define these ones.

rajiv448 replied:
Done. Pushed in latest updates

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.h:
```diff
@@ -0,0 +1,40 @@
+  public:
+  xrt_device_ftbl     device;
+
+  static xrt_ftbl& get_instance()
+  {
+    static xrt_ftbl instance;
+    return instance;
+  }
```
hmm, will this work as expected in a header file?  What if this header is included in multiple compilation units.?  Maybe the linker will handle it, but I would implement get_instance() in capture.cpp rather than in header.

Also, I don't think xrt_ftbl should be a class.  This is a singleton, a namespace would suffice with a method that returns what you need.

In header:
```
namespace xrt_ftbl {

xrt_device_ftbl
get_tbl();

}
```

in cpp:
```
namespace xrt_core::tools::xbtracer {
...

namespace xrt_ftbl {
xrt_device_ftbl* // or reference 
get_tbl()
{
   static xrt_device_ftbl device;
   return &device;
}
} // xrt_ftbl

....
} // xrt_core::tools::xbtracer
```

rajiv448 replied:
I have made changes related to defining get_instance(). But has still kept xrt_ftbl as class. Please see if this would alright or suggest if I should indeed change it to namespace.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.h:
```diff
@@ -0,0 +1,40 @@
+
+#pragma once
+
+#include "experimental/xrt_elf.h"
+#include "experimental/xrt_hw_context.h"
+#include "experimental/xrt_ip.h"
+#include "experimental/xrt_module.h"
+#include "experimental/xrt_system.h"
+#include "experimental/xrt_xclbin.h"
+#include "xrt/xrt_bo.h"
+#include "xrt/xrt_device.h"
+#include "xrt/xrt_graph.h"
+#include "xrt/xrt_kernel.h"
```
Include-what-you-use.   Never include more headers than what is needed to compile the file (in this case capture.h) into an object file.   The source (.cpp) files include what these headers if they use them.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp:
```diff
@@ -0,0 +1,334 @@
+#include <mutex>
+#include <array>
+
+#ifdef _WIN32
+  #include <windows.h>
+#else
+  #include <pthread.h>
+  #include <unistd.h>
+#endif /* #ifdef _WIN32 */
```
Nit.
```suggestion
#ifdef _WIN32
# include <windows.h>
#else
# include <pthread.h>
# include <unistd.h>
#endif /* #ifdef _WIN32 */
```

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp:
```diff
@@ -0,0 +1,334 @@
+constexpr unsigned int str_sz_m = 128;
+constexpr unsigned int str_sz_l = 256;
+constexpr unsigned int fw_9 = 9;
+#ifdef _WIN32
+constexpr std::string_view path_separator = "\\";
+#else
+constexpr std::string_view path_separator = "/";
+#endif /* #ifdef _WIN32 */
```
There is too much ifdef this and that.   It is much better to have `detail/windows/` and `detail/linux/` directory to define the differences.   This source file (logger.cpp) should just include `#include detail/logger.h` which in turn ifdef windows and linux to include `detail/windows/logger.h` and `detail/linux/logger.h`  .  Hide the platform details in appropriate header files that are included appropriately.  See `core/common/detail` for examples.

rajiv448 replied:
Done, not as cleanly as in the reference. Please review and let me know if you if this would be sufficient.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp:
```diff
@@ -0,0 +1,334 @@
+std::string logger::os_name_ver()
+{
+  std::string pretty_name;
+#ifdef _WIN32
```
Move to details.

rajiv448 replied:
Done

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp:
```diff
@@ -0,0 +1,334 @@
+ * */
+logger::~logger()
+{
+  std::string time_fmt_str="";
+  struct timespec time_now{};
+
+  #ifndef _WIN32
+  clock_gettime (CLOCK_REALTIME, &time_now);
+  #endif
+
+  read_time_now_fmt(time_fmt_str, time_now);
+
+  fp << "|END|" << time_fmt_str << "." << std::setfill('0') << std::setw(fw_9)
+     << time_now.tv_nsec << "|\n";
+
+  fp_bin.close();
+  fp.close();
```
Almost every single line in this dtor can throw.   Dtors are now allowed to throw, do try/catch .

rajiv448 replied:
Wil do it in next update.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp:
```diff
@@ -0,0 +1,334 @@
+  struct timespec time_now{};
+  std::stringstream ss;
+
+#ifdef _WIN32
+  DWORD tid = GetCurrentThreadId();
+  read_time_now(time_now);
+#else
+  pthread_t tid = pthread_self();
+  clock_gettime (CLOCK_REALTIME, &time_now);
+#endif /* #ifdef _WIN32 */
```
`auto tid = detail::current_thread_id()`

Or why use pthread?  Why not std::thread?

rajiv448 replied:
Done. used std::thread

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.h:
```diff
@@ -0,0 +1,323 @@
+/*
+ * enumration to identify the log type (Entry/Exit)
+ * */
+enum class trace_type {
+  TRACE_TYPE_ENTRY,
+  TRACE_TYPE_EXIT,
+  TRACE_TYPE_INVALID
+};
```
Do not capitalize enumerators in class enum and don't bother prefixing them.  The enumerators here should simply be: `entry`, `exit`, `invalid`.

rajiv448 replied:
Sure, I have made changes and submit it in upcoming update

stsoe reviewed (changes requested):
There is too much ifdef windows and linux cluttering the code. 
Please check comments and consider refactoring into some details/

rajiv448 committed:
Added interception support for linux platform

few APIs from xrt::device has been instrumented in this commit
other required infrastructure for linux platform has been fully
commited.

rajiv448 committed:
Added logger module.

This module provide infra to trace the APIs and data.
Traces(.txt and .bin) are written into seperate directory based
on execution timestamp.

rajiv448 committed:
Added interception support for windows platform

rajiv448 committed:
Added instrumentation wrapper for other APIs

extended instrumentation to other APIs planned for tracing

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+link_directories($ENV{XILINX_XRT}/lib)
+
+set(SRCS
+  capture.cpp
+  logger.cpp
+  xrt_device_inst.cpp
+  xrt_kernel_inst.cpp
+  xrt_bo_inst.cpp
+  xrt_hw_context_inst.cpp
+  xrt_xclbin_inst.cpp
+)
+
+set (LIBS xrt_coreutil)
```
My recommendation is to not use these sort of variables, but instead list the files explicitly in the targets.

rajiv448 replied:
changed as recommended, please check the latest update

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+
+  public:
+  router(const router&);
+  auto operator = (const router&) -> router&;
+  router(router && other) noexcept;
+  auto operator = (router&& other) noexcept -> router&;
+  auto load_func_addr() -> int;
+  auto load_symbols() -> int;
```
In general `-> return_type` doesn't read as well as listing the return_type in place of auto.   There are only very few cases where trailing return_type is needed and when not, the syntax is not wide adopted in C++.

rajiv448 replied:
Was handled in 9/25 update

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+ * with address of the functions from original
+ * library.
+ */
+int router::load_func_addr()
```
My preference on error codes is not to use them and instead throw on error.  Don't know for sure if exceptions fit here?

rajiv448 replied:
Sure, we have made the changes. Please check the lasted updated (9/26)

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+  /**
+   * Get Function address's which are of interest and ignore others.
+   */
+  for (auto& it : func_mangled)
```
Suggest to use structured binding:
`for (auto& [v1, v2] : func_manged)`
Replace `v1` `v2` with better names.

rajiv448 replied:
Updated as suggested in latest update, please check.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+{
+  constexpr unsigned int symbol_len = 1024;
+
+  path = find_library_path();
```
Searching for path, I assume it is a data member of `router`.   I have done what you have done, but always regretted it, I am now quite consistent in using `m_<var>` for data members of classes.

rajiv448 replied:
Changed path to m_path as suggested.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+ * This function will read mangled API's from library and perform
+ * Demangling operation.
+ */
+int router::load_symbols()
```
ditto on return error code here.

Also this is quite a long functions, any possibility of refactoring to make it more readable?

rajiv448 replied:
These comments were folded and overlooked them, my bad. I have address these, please check.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+    }
+  }
+
+  if (symtab_section == nullptr)
```
No common to compare against nullptr.
```suggestion
  if (!symtab_section)
```

rajiv448 replied:
Done

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+      std::string demangled_and_conditioned_str =
+          find_and_replace_all(demangled_str, replacements);
+      return demangled_and_conditioned_str;
+    } else {
```
align with `if` ?

rajiv448 replied:
Done

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+  }
+}  // namespace xrt::tools::xbtracer
+
+using namespace xrt::tools::xbtracer;
```
Using does often make code more difficult to read.  You may want a namespace alias instead?
```suggestion
 namespace xtx = xrt::tools::xbtracer;
```

rajiv448 replied:
Done as suggested, please check.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.cpp:
```diff
@@ -0,0 +1,617 @@
+// Iterate through the IDT for all table entry corresponding to xrt_coreutil.dll
+// and replace the function pointer in first_thunk by looking for the same name
+// into the xrt_capture.dll for the same name.
+int idt_fixup(void* dummy)
```
return code and rather long function :-)

rajiv448 replied:
made the function return void. Also added 2 subfunctions to this method.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/capture.h:
```diff
@@ -0,0 +1,53 @@
+
+#pragma once
+
+#include <mutex>
+#include "experimental/xrt_ext.h"
+#include "experimental/xrt_hw_context.h"
+#include "experimental/xrt_xclbin.h"
+#include "xrt/xrt_bo.h"
+#include "xrt/xrt_device.h"
+#include "xrt/xrt_kernel.h"
```
I don't see any of these used within this header file.  Always include-what-you-use and where-you-use.

rajiv448 replied:
moved to respective internal headers where they are needed.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/detail/windows/logger.h:
```diff
@@ -0,0 +1,88 @@
+  return bt;
+}
+
+std::string get_env(std::string key)
```
It's not for free to pass std::string by value; use const reference?

rajiv448 replied:
Done

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/detail/windows/logger.h:
```diff
@@ -0,0 +1,88 @@
+
+std::string get_env(std::string key)
+{
+  char *val = nullptr;
+  size_t len = 0;
+  errno_t err = _dupenv_s(&val, &len, key.c_str());
+  if (err != 0 || val == nullptr)
+    return std::string();
+  std::string result(val);
+  free(val); // Free the memory allocated by _dupenv_s
+  return result;
```
std::string constructor may throw in which case `val` is leaked, in other words, this function is not exception safe.

rajiv448 replied:
Done, please check. Let me know if any further change needed.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/detail/windows/logger.h:
```diff
@@ -0,0 +1,88 @@
+
+std::string wide_to_string(const std::wstring& wstr)
+{
+  if (wstr.empty()) return std::string();
```
Always put stmt on new line otherwise debugging and setting breaks points is difficult,
```suggestion
  if (wstr.empty()) 
    return std::string();
```

rajiv448 replied:
agree, updated as suggested.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/detail/windows/logger.h:
```diff
@@ -0,0 +1,88 @@
+
+std::string get_os_name_ver()
+{
+  std::string pretty_name;
```
Move to where used (line 53).

rajiv448 replied:
Moved as suggested in latest update, please check.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/detail/windows/logger.h:
```diff
@@ -0,0 +1,88 @@
+      = (RtlGetVersionPtr)GetProcAddress(GetModuleHandleW(L"ntdll.dll"),
+                                         "RtlGetVersion");
+
+  if (RtlGetVersion != nullptr)
```
```suggestion
  if (RtlGetVersion)
```

rajiv448 replied:
Updated, Please check

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.cpp:
```diff
@@ -0,0 +1,234 @@
+{
+  bool run = true;
+
+  while(run)
```
```suggestion
  while (run)
```

rajiv448 replied:
Updated, please check.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/lib/logger.h:
```diff
@@ -0,0 +1,392 @@
+  logger();
+
+  public:
+  bool get_inst_debug()
```
Just noticing now.  Make sure you declare member functions `const` when they can be.
```suggestion
  bool get_inst_debug() const
```

rajiv448 replied:
Updated as suggested, please check.

stsoe reviewed (changes requested):
Some suggestions.

rajiv448 commented:
@stsoe - Thanks for your previous reviews and input. Updated by incorporating all changes as suggested. Please re-review

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -118,6 +139,65 @@ void log_d(const Args&... args)
+#endif
+}
+
+std::string get_env(std::string key)
```
```suggestion
std::string get_env(const std::string& key)
```

rajiv448 replied:
Done

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -118,6 +139,65 @@ void log_d(const Args&... args)
+std::string get_env(std::string key)
+{
+#if _WIN32
+  char *val = nullptr;
+  size_t len = 0;
+  errno_t err = _dupenv_s(&val, &len, key.c_str());
+  if (err != 0 || val == nullptr)
+    return std::string();
+  std::string result(val);
+  free(val); // Free the memory allocated by _dupenv_s
```
This is not exception safe. std::string ctor may throw.   This _dupenv_s business is very unfortunate, what does MSFT recommend in order to make this exception safe?   I think you have to create a unique_ptr with val immediately after _dupenv_s , the unique_ptr should call `free` when going out of scope.   Horrible mess :-(

rajiv448 replied:
Done, please check if it looks alright.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -118,6 +139,65 @@ void log_d(const Args&... args)
+std::string find_file(const std::string& path, const std::string& file)
+{
+  std::istringstream path_stream(path);
+  std::string l_path, full_path;
+  while (std::getline(path_stream, l_path, path_delimiter))
+  {
+    full_path = l_path + std::string(path_separator) + file;
```
```suggestion
  std::string l_path;
  while (std::getline(path_stream, l_path, path_delimiter))
  {
    auto full_path = l_path + std::string(path_separator) + file;
```

rajiv448 replied:
Done

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -282,28 +524,89 @@ int create_child_proc_as_suspended(launcher& app)
     log_e("Waiting for child process failed");
-    return -2;
+    retcode = -2;
+    goto l_exit;
```
This is not right.   I understand we clean up but it is not cool, and definitely not exception safe.  I don't have `log_e` handy but it's probably not nothrow ?  Are you sure all exit paths of this function will actually reach `l_exit:` ?

rajiv448 replied:
Although log_e doesn't throw and all path were reaching l_exit:, to remove goto refactored the code.

stsoe commented on src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp:
```diff
@@ -282,28 +524,89 @@ int create_child_proc_as_suspended(launcher& app)
+        (LPTHREAD_START_ROUTINE)GetProcAddress(app.m_hlib, "idt_fixup");
+
+    if (app.m_idt_fixup == nullptr)
+      return FALSE;
```
what's wrong with `false` and 'true' ?

rajiv448 replied:
Done

stsoe reviewed (changes requested):
Left more comments, but I don't see any changes compared to last time I reviewed?

rajiv448 commented:
> Left more comments, but I don't see any changes compared to last time I reviewed?

I will address new comments and update the PR. Please let me know if there are more input. Following are the changes related to last set of review comments. 
https://github.com/rajiv448/XRT/compare/703f6c909e07c125bc593274c50887172cc5f1b8..dcde59d5a849a4022d6f299cddde9f0cf3d18f24

Please let me know, if I am overlooking or missing anything.

rajiv448 commented:
> > Left more comments, but I don't see any changes compared to last time I reviewed?
> 
> I will address new comments and update the PR. Please let me know if there are more input. Following are the changes related to last set of review comments. https://github.com/rajiv448/XRT/compare/703f6c909e07c125bc593274c50887172cc5f1b8..dcde59d5a849a4022d6f299cddde9f0cf3d18f24
> 
> Please let me know, if I am overlooking or missing anything.

I realized today that because of folded comment, I overlooked few of the input. I have address all of them including new comments. Please review and let us know your input.

rajiv448 commented:
As discussed in the meeting today, we are merging these changes.
